src/core/thread/link_metrics.cpp

```diff
@@ -44,39 +44,177 @@
 
 namespace ot {
 
+void LinkMetricsSeriesInfo::Init(uint8_t              aSeriesId,
+                                 const SeriesFlags &  aSeriesFlags,
+                                 const otLinkMetrics &aLinkMetricsFlags)
+{
+    mSeriesId                = aSeriesId;
+    mSeriesFlags             = aSeriesFlags;
+    mLinkMetrics.mPduCount   = aLinkMetricsFlags.mPduCount;
+    mLinkMetrics.mLqi        = aLinkMetricsFlags.mLqi;
+    mLinkMetrics.mLinkMargin = aLinkMetricsFlags.mLinkMargin;
+    mLinkMetrics.mRssi       = aLinkMetricsFlags.mRssi;
+    mRssAverager.Reset();
+    mLqiAverager.Reset();
+    mPduCount = 0;
+}
+
+void LinkMetricsSeriesInfo::AggregateLinkMetrics(uint8_t aFrameType, uint8_t aLqi, int8_t aRss)
+{
+    if (IsFrameTypeMatch(aFrameType))
+    {
+        mPduCount++;
+        mLqiAverager.Add(aLqi);
+        IgnoreError(mRssAverager.Add(aRss));
+    }
+}
+
+void LinkMetricsSeriesInfo::GetLinkMetrics(otLinkMetrics &aLinkMetrics) const
+{
+    aLinkMetrics.mPduCount   = mLinkMetrics.mPduCount;
+    aLinkMetrics.mLqi        = mLinkMetrics.mLqi;
+    aLinkMetrics.mLinkMargin = mLinkMetrics.mLinkMargin;
+    aLinkMetrics.mRssi       = mLinkMetrics.mRssi;
+}
+
+bool LinkMetricsSeriesInfo::IsFrameTypeMatch(uint8_t aFrameType) const
+{
+    bool match = false;
+
+    switch (aFrameType)
+    {
+    case kSeriesTypeLinkProbe:
+        VerifyOrExit(!mSeriesFlags.IsMacDataFlagSet()); // Ignore this when Mac Data is accounted
+        match = mSeriesFlags.IsLinkProbeFlagSet();
```

Comment:
1.2 spec 4.11.3.4.4.4 says
> Bit 0: Track metrics of forward MAC data frames for MLE Link Probes that contain a Link Probe TLV with a ‘Series ID’ field that matches the ‘Forward Series ID’ of this series.
 
> Bit 1: Track metrics of forward data frames for all Link Layer data frames. If this bit is selected, Bit 0 SHALL be ignored.

My understanding is 
- if `IsMacDataFlagSet()` is false,  match should not be true unless the `Serial ID` in `MLE Link Probe` matches the one in the `LinkMetricsSeriesInfo`; 
- if `IsMacDataFlagSet()` is true, any MLE Link Probe should be valid MAC Data frame, regardless of the `Series ID` inside.

However here I didn't see check on Series ID match, this code block ignores MLE link probe, not the expecting behavior to ignore `bit 0` of the series flags.

Reply:
Updated to add match for `SeriesID` for Link Probe.

Reply:
@librasungirl

If `IsMacDataFlagSet` is true, this message would be accounted as a mac data frame. `VerifyOrExit(!mSeriesFlags.IsMacDataFlagSet());` is used to avoid duplication.

For example, when `IsMacDataFlagSet` is `true` and we receive a Link Probe message. The Link Probe message would be accounted first at `AggregateLinkMetrics` in `Mac::HandleReceivedFrame` because it is itself a link layer data frame. And then accounted again at `AggregateLinkMetrics` in `Mle::HandleLinkProbe`. I did this because Link Probe are not link layer messages so I cannot directly count it in Mac. If `IsMacDataFlagSet` is true, Link Probe would be accounted in Mac layer and won't be accounted in Mle. If `IsMacDataFlagSet` is false and `IsLinkProbeFlagSet()` is `true`, Link Probe would be accounted in Mle.

Reply:
yes, you are right.